Remove RTTI from jit/type.h

bwasti opened the pull request:
Summary: RTTI can't be used on android so this is needed

Differential Revision: D12914402

bwasti commented:
@pytorchbot retest this please

Yangqing reviewed (approved):
Hmm, this looks a bit manual, but I assume that if we want to keep cast safety then we have to do this anyway - LGTM

bwasti commented:
@Yangqing the boiler plate is pretty much the only way to get rid of RTTI.   The only change over the LLVM style is to invert the "classof" structure and make that function pure virtual (rather than static), which gives compile time checks for any new types.  It makes the usage a bit different, but for such a limited use-case I think it's fine

bwasti committed:
Remove RTTI from jit/type.h (#13591)

Summary:
RTTI can't be used on android so this is needed
Pull Request resolved: https://github.com/pytorch/pytorch/pull/13591

Differential Revision: D12914402

Pulled By: bwasti

fbshipit-source-id: 60c39b7f8c39836e20ce98609118f76952a32402
